src/main/scala/beam/agentsim/agents/vehicles/VehicleEnergy.scala

```diff
@@ -0,0 +1,82 @@
+
+  type ConsumptionRateFilter = Map[Range, Map[Range, Map[Range, Float]]] //speed->(gradePercent->(numberOfLanes->rate))
+  private lazy val consumptionRateFilter = loadConsumptionRatesFromCSV
+
+  def getRateUsing(travelData: TravelData, fallBack: => Float): Float = {
+    val TravelData(speed, gradePercent, numberOfLanes) = travelData
+    //Future performance improvement could be to better index the bins so could fuzzily jump straight to it
+    //instead of having to iterate
+    val filteredRates = consumptionRateFilter
+      .collect{case (speedBin, restOfFilter) if speedBin.has(speed) => restOfFilter}.flatten
+      .collect{case (gradePercentBin, restOfFilter) if gradePercentBin.has(gradePercent) => restOfFilter}.flatten
+      .collect{case (numberOfLanesBin, rate) if numberOfLanesBin.has(numberOfLanes) => rate}
+
+    val ratesSize = filteredRates.size
+    if(ratesSize > 1) log.warn("More than one ({}) rate was found using {}. " +
+      "The first will be used, but the data should be reviewed for range overlap.", ratesSize, travelData)
+    filteredRates.headOption.getOrElse(fallBack)
+  }
+
+  private def loadConsumptionRatesFromCSV: ConsumptionRateFilter = {
+    val currentRateFilter = mutable.Map.empty[Range, mutable.Map[Range, mutable.Map[Range, Float]]]
+    recordIterator.foreach(csvRecord => {
+      val speedBin = Range(csvRecord.getString(0).replace(",",":").replace(" ",""))
+      val gradePercentBin = Range(csvRecord.getString(1).replace(",",":").replace(" ",""))
+      val numberOfLanesBin = Range(csvRecord.getString(2).replace(",",":").replace(" ",""))
+      val rateStr = csvRecord.getString(5)
+      println(csvRecord.getValues.toList)
+      println(rateStr)
+      val rate = rateStr.toFloat
+      currentRateFilter.get(speedBin) match {
+        case Some(gradePercentFilter) => {
+          gradePercentFilter.get(gradePercentBin) match {
+            case Some(numberOfLanesFilter) => {
+              numberOfLanesFilter.get(numberOfLanesBin) match {
+                case Some(initialRate) =>
+                  println(initialRate)
+                  println(rate)
+                  log.error("Two rates found for the same bin combination: " +
+                    "Speed Bin = {}; Grade Percent Bin = {}; Number of Lanes Bin = {}. " +
+                    "Keeping initial rate of {} and ignoring new rate of {}.",
+                    speedBin.toString, gradePercentBin.toString, numberOfLanesBin.toString, initialRate.toString, rate.toString)
+                case None => numberOfLanesFilter += numberOfLanesBin -> rate
+              }
+            }
+            case None => gradePercentFilter += gradePercentBin -> mutable.Map(numberOfLanesBin -> rate)
+          }
+        }
+        case None => currentRateFilter += speedBin -> mutable.Map(gradePercentBin -> mutable.Map(numberOfLanesBin -> rate))
+      }
+    })
+    println(currentRateFilter)
+    currentRateFilter.toMap.map{
```

Comment:
This might not be necessary if the return type is set to immutable maps. But since this should only be done once and cached, then I don't see a need to remove it for now.